### PR TITLE
[improve](txn insert) Txn load support cloud mode

### PR DESCRIPTION
**Pull request by mymeiyi**
### Purpose

The user doc: https://doris.apache.org/zh-CN/docs/dev/data-operate/import/transaction-load-manual

We have supported insert into select(https://github.com/apache/doris/pull/31666), update(https://github.com/apache/doris/pull/33034) and delete(https://github.com/apache/doris/pull/33100) in transaction load.

https://github.com/apache/doris/pull/32980 implements one txn write to one partition more than one rowsets.

This pr implements to cloud mode of https://github.com/apache/doris/pull/32980

#### sub_txn_id

see https://github.com/apache/doris/pull/32980

#### Meta service supports commit txn

This process is generally the same as commit_txn, the difference is that he partitions version will plus 1 in multi sub txns.

One example:
Suppose the table, partition, tablet and version info is:
```
--------------------------------------------
| table | partition | tablet    | version |
--------------------------------------------
| t1    | t1_p1     | t1_p1.1   | 1       |
| t1    | t1_p1     | t1_p1.2   | 1       |
| t1    | t1_p2     | t1_p2.1   | 2       |
| t2    | t2_p3     | t2_p3.1   | 3       |
| t2    | t2_p4     | t2_p4.1   | 4       |
--------------------------------------------
```

Now we commit a txn with 3 sub txns and the tablets are:
 *  sub_txn1: t1_p1.1, t1_p1.2, t1_p2.1
 *  sub_txn2: t2_p3.1
 *  sub_txn3: t1_p1.1, t1_p1.2

When commit, the partitions version will be:
 *  sub_txn1: t1_p1(1 -> 2), t1_p2(2 -> 3)
 *  sub_txn2: t2_p3(3 -> 4)
 *  sub_txn3: t1_p1(2 -> 3)

After commit, the partitions version will be:
 *  t1: t1_p1(3), t1_p2(3)
 *  t2: t2_p3(4), t2_p4(4)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mymeiyi** [comment]
run buildall

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 1]
```diff
@@ -1235,6 +1239,551 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
     response->mutable_txn_info()->CopyFrom(txn_info);
 } // end commit_txn
 
+void MetaServiceImpl::commit_txn_with_sub_txn(::google::protobuf::RpcController* controller,
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1242]: add comment to describe what does this rpc do?

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by mymeiyi** [comment]
run cloud_p0 cloud_ut

**Comment by mymeiyi** [comment]
run cloud_p0

**Review by dataroaring — DISMISSED** [review]
LGTM

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 2]
```diff
@@ -1235,161 +1239,735 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
...
+        new_versions[version_keys[i]] = version;
+        LOG(INFO) << "xxx get partition_version_key=" << hex(version_keys[i])
+                  << " version:" << version << " txn_id=" << txn_id;
+    }
+    version_keys.clear();
+    version_values.clear();
+
+    std::vector<std::pair<std::string, std::string>> rowsets;
+    std::unordered_map<int64_t, TabletStats> tablet_stats; // tablet_id -> stats
+    for (const auto& sub_txn_info : sub_txn_infos) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1562]: Add enough UT to test the rowset meta KV versions continuity

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 3]
```diff
@@ -1235,161 +1239,735 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
...
+        code = MetaServiceCode::TXN_ALREADY_ABORTED;
+        ss << "transaction is already aborted: db_id=" << db_id << " txn_id=" << txn_id;
+        msg = ss.str();
+        LOG(WARNING) << msg;
+        return;
+    }
+
+    if (txn_info.status() == TxnStatusPB::TXN_STATUS_VISIBLE) {
+        code = MetaServiceCode::TXN_ALREADY_VISIBLE;
+        if (request->has_is_2pc() && request->is_2pc()) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1443]: do not support 2PC

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 4]
```diff
@@ -1235,161 +1239,735 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
...
+    if (!recycle_pb.SerializeToString(&recycle_val)) {
+        code = MetaServiceCode::PROTOBUF_SERIALIZE_ERR;
+        ss << "failed to serialize recycle_pb, txn_id=" << txn_id;
+        msg = ss.str();
+        return;
+    }
+    txn->put(recycle_key, recycle_val);
+
+    LOG(INFO) << "xxx commit_txn put recycle_key key=" << hex(recycle_key) << " txn_id=" << txn_id;
+    LOG(INFO) << "commit_txn put_size=" << txn->put_bytes() << " del_size=" << txn->delete_bytes()
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1784]: one LOG

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by dataroaring** [comment]
run buildall

File: gensrc/proto/cloud.proto [hunk 5]
```diff
@@ -646,6 +648,15 @@ message CommitTxnRequest {
...
     repeated int64 base_tablet_ids= 7; // all tablet from base tables (excluding mv)
+    // for transaction load
+    optional bool is_txn_load = 9;
+    repeated SubTxnInfo sub_txn_infos = 10;
+}
+
+message SubTxnInfo {
+    optional int64 sub_txn_id = 1;
+    optional int64 table_id = 2;
+    repeated int64 base_tablet_ids= 3;
```
[the remark this answers is not in the document]
Comment (mymeiyi) [reply gensrc/proto/cloud.proto:659]: The Coordinator::updateCommitInfos ensure it

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run cloud_ut cloud_p1

**Comment by mymeiyi** [comment]
run cloud_p1

**Comment by mymeiyi** [comment]
run cloud_p1

**Comment by mymeiyi** [comment]
run cloud_p0

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run p0

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by mymeiyi** [comment]
run fe_ut

**Comment by mymeiyi** [comment]
run performance

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run cloud_p0

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by mymeiyi** [comment]
run buildall

**Comment by mymeiyi** [comment]
run cloud_ut

**Comment by mymeiyi** [comment]
run p0

**Comment by mymeiyi** [comment]
run cloudut

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**